actpool/blobstore.go

```diff
@@ -0,0 +1,245 @@
+package actpool
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/holiman/billy"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/pkg/log"
+)
+
+const (
+	// blobSize is the protocol constrained byte size of a single blob in a
+	// transaction. There can be multiple of these embedded into a single tx.
+	blobSize = params.BlobTxFieldElementsPerBlob * params.BlobTxBytesPerFieldElement
+
+	// maxBlobsPerTransaction is the maximum number of blobs a single transaction
+	// is allowed to contain. Whilst the spec states it's unlimited, the block
+	// data slots are protocol bound, which implicitly also limit this.
+	maxBlobsPerTransaction = params.MaxBlobGasPerBlock / params.BlobTxBlobGasPerBlob
+
+	// txAvgSize is an approximate byte size of a transaction metadata to avoid
+	// tiny overflows causing all txs to move a shelf higher, wasting disk space.
+	txAvgSize = 4 * 1024
+
+	// txMaxSize is the maximum size a single transaction can have, outside
+	// the included blobs. Since blob transactions are pulled instead of pushed,
+	// and only a small metadata is kept in ram, the rest is on disk, there is
+	// no critical limit that should be enforced. Still, capping it to some sane
+	// limit can never hurt.
+	txMaxSize = 1024 * 1024
+)
+
+type (
+	blobStore struct {
+		config blobStoreConfig // Configuration for the blob store
+
+		store  billy.Database // Persistent data store for the tx
+		stored uint64         // Useful data size of all transactions on disk
+
+		lookup map[hash.Hash256]uint64 // Lookup table mapping hashes to tx billy entries
+		lock   sync.RWMutex            // Mutex protecting the store
+
+		encode encodeAction // Encoder for the tx
+		decode decodeAction // Decoder for the tx
+	}
+	blobStoreConfig struct {
+		Datadir string `yaml:"datadir"` // Data directory containing the currently executable blobs
+		Datacap uint64 `yaml:"datacap"` // Soft-cap of database storage (hard cap is larger due to overhead)
+	}
+
+	onAction     func(selp *action.SealedEnvelope) error
+	encodeAction func(selp *action.SealedEnvelope) ([]byte, error)
+	decodeAction func([]byte) (*action.SealedEnvelope, error)
+)
+
+var (
+	errBlobNotFound = fmt.Errorf("blob not found")
+)
+
+var defaultBlobStoreConfig = blobStoreConfig{
+	Datadir: "blobpool",
+	Datacap: 10 * 1024 * 1024 * 1024,
```

Comment:
evmNetworkID should be one of the config value?

Reply:
not required, it's inner part of decode

Reply:
yes it's part of decode, think about we start the pool on testnet vs. mainnet, there needs to be a way to pass in `evmNetworkID`?

Reply:
can ref to here for the usage: https://github.com/iotexproject/iotex-core/pull/4366/files#diff-c65dd5a657c73eb89458c69a82a25fc0993655650b3ad477136955c31e50e2bbR249

Reply:
I see, so pass in the `decode` func which takes care of the evmNetworkID